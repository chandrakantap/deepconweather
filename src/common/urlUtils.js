export function getQueryParams(searchString = "") {
  return searchString
    .slice(1)
    .split("&")
    .map((p) => p.split("="))
    .reduce((obj, [key, value]) => ({ ...obj, [key]: value }), {});
}
