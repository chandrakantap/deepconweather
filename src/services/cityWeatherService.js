import { getCityWeather } from "./weatherStackApi";
import userNoteService from "./userNoteService";
import topCitiesByPopulation from "./topCitiesByPopulation";

export const LIST_PAGE_CITIES_SK = "SK_LIST_PAGE_CITIES";

function getCityList() {
  const listPageCitiesString = localStorage.getItem(LIST_PAGE_CITIES_SK);
  let listPageCities;

  if (!listPageCitiesString) {
    listPageCities = topCitiesByPopulation.slice(0, 15).map((city) => ({
      ...city,
      isFavourite: false,
      current: { weather_descriptions: ["-"] },
      id: `${city.name}_${city.region}_${city.country}`
        .toLocaleUpperCase()
        .replace(/[ ]*/g, ""),
    }));
    localStorage.setItem(LIST_PAGE_CITIES_SK, JSON.stringify(listPageCities));
  } else {
    listPageCities = JSON.parse(listPageCitiesString);
  }
  return listPageCities;
}

async function refreshCityListWeather() {
  const cities = getCityList();
  const refreshedCityList = await Promise.all(
    cities.map((city) => getCityWeatherDetail(city))
  );
  localStorage.setItem(LIST_PAGE_CITIES_SK, JSON.stringify(refreshedCityList));
  return refreshedCityList;
}

async function getCityWeatherDetail(city) {
  const query = `${city.name},${city.region},${city.country}`;
  const { current } = await getCityWeather(query);
  return { ...city, current };
}

function toggleFavourite(cityId) {
  const listPageCities = getCityList();
  const updatedCityData = listPageCities.map((city) => {
    return city.id === cityId
      ? { ...city, isFavourite: !city.isFavourite }
      : city;
  });
  localStorage.setItem(LIST_PAGE_CITIES_SK, JSON.stringify(updatedCityData));
}

function removeCity(cityId) {
  const listPageCities = getCityList();
  const updatedCityData = listPageCities.filter((city) => city.id !== cityId);
  localStorage.setItem(LIST_PAGE_CITIES_SK, JSON.stringify(updatedCityData));
  userNoteService.removeUserNoteForCity(cityId);
}

function addCity(newCity) {
  const currentCities = getCityList().filter((city) => city.id !== newCity.id);
  localStorage.setItem(
    LIST_PAGE_CITIES_SK,
    JSON.stringify([...currentCities, newCity])
  );
}

export default {
  getCityList,
  refreshCityListWeather,
  getCityWeatherDetail,
  toggleFavourite,
  addCity,
  removeCity,
};
