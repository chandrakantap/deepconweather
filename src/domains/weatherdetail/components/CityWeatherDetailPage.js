import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { MdHome } from "react-icons/md";
import IconButton from "common/ui/IconButton";
import Throbber from "common/ui/Throbber";
import {
  loadCityWeatherDetailAction,
  clearStateAction,
} from "domains/weatherdetail/store/cityWeatherDetailPageActions";
import CityTemperatureDetail from "./CityTemperatureDetail";
import WeatherAttributes from "./WeatherAttributes";
import UserNotes from "domains/userNotes/components/UserNotes";
import { isDetailsLoadedSelector } from "domains/weatherdetail/store/cityWeatherDetailPageSelectors";
import { getQueryParams } from "common/urlUtils";
import styles from "./CityWeatherDetailPage.module.css";

function CityWeatherDetailPage(props) {
  const dispatch = useDispatch();
  const isLoaded = useSelector(isDetailsLoadedSelector);

  useEffect(() => {
    const { cityId, cityName, region, country } = getQueryParams(
      decodeURI(props.location.search)
    );
    dispatch(
      loadCityWeatherDetailAction({ cityId, cityName, region, country })
    );
    return () => dispatch(clearStateAction());
  }, [dispatch, props.location.search]);

  const onClickBack = () => {
    props.history.push("/list");
  };
  if (!isLoaded) {
    return (
      <section className={styles.detailPageRoot}>
        <Throbber />
      </section>
    );
  }
  return (
    <section className={styles.detailPageRoot}>
      <div className={styles.backButton}>
        <IconButton onClick={onClickBack}>
          <MdHome /> &nbsp;&nbsp;Home
        </IconButton>
      </div>
      <CityTemperatureDetail />
      <div className={styles.lowerSection}>
        <WeatherAttributes />
        <div />
        <UserNotes />
      </div>
    </section>
  );
}

export default CityWeatherDetailPage;
