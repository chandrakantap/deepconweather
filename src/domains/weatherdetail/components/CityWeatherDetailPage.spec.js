import React from "react";
import {
  fireEvent,
  waitForDomChange,
  screen,
  cleanup,
  act,
} from "@testing-library/react";
import { render } from "common/test-utils";
import cityWeatherService from "services/cityWeatherService";
import CityWeatherDetailPage from "./CityWeatherDetailPage";

const cityData = {
  name: "Lobong",
  region: "Pangasinan",
  country: "Philippines",
  id: "LOBONG_PANGASINAN_PHILIPPINES",
  isFavourite: true,
  current: {
    temperature: 24,
    weather_descriptions: ["Moderate or heavy rain shower"],
    wind_dir: "E",
    pressure: 1006,
    precip: 0.9,
    humidity: 89,
    cloudcover: 37,
    feelslike: 26,
    uv_index: 1,
    visibility: 9,
    is_day: "no",
  },
};

describe("CityWeatherDetailPage", () => {
  test("should renders properly when loading", async () => {
    const location = {
      search:
        "?cityId=LOBONG_PANGASINAN_PHILIPPINES&cityName=Lobong&region=Pangasinan&country=Philippines",
    };
    const { container } = render(<CityWeatherDetailPage location={location} />);
    expect(container).toMatchSnapshot();
  });
  test("should renders properly after data loaded", async () => {
    const location = {
      search:
        "?cityId=LOBONG_PANGASINAN_PHILIPPINES&cityName=Lobong&region=Pangasinan&country=Philippines",
    };
    jest
      .spyOn(cityWeatherService, "getCityWeatherDetail")
      .mockResolvedValueOnce(cityData);
    render(<CityWeatherDetailPage location={location} />);
    await waitForDomChange();
    expect(screen.getByText("Pangasinan, Philippines")).toBeTruthy();
  });
  test("should toggle city as favourite", async () => {
    const location = {
      search:
        "?cityId=LOBONG_PANGASINAN_PHILIPPINES&cityName=Lobong&region=Pangasinan&country=Philippines",
    };
    jest
      .spyOn(cityWeatherService, "getCityWeatherDetail")
      .mockResolvedValueOnce(cityData);
    const { container } = render(<CityWeatherDetailPage location={location} />);
    await waitForDomChange();
    fireEvent.click(screen.getByTestId("DetailPage_favBtn"));
    expect(container).toMatchSnapshot();
  });
});
