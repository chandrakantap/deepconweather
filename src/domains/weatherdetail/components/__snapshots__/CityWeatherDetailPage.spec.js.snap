// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityWeatherDetailPage should renders properly when loading 1`] = `
<div>
  <section
    class="detailPageRoot"
  >
    <div
      class="throbber"
    >
      <div
        class="throbberMask"
      />
      <div
        class="loaderContainer"
      >
        <div
          class="loader"
        />
      </div>
    </div>
  </section>
</div>
`;

exports[`CityWeatherDetailPage should toggle city as favourite 1`] = `
<div>
  <section
    class="detailPageRoot"
  >
    <div
      class="backButton"
    >
      <button
        class="iconButton"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
          />
        </svg>
           Home
      </button>
    </div>
    <div
      class="mainDetails"
    >
      <div
        class="cityName"
      >
        <h1>
          Lobong
          <button
            class="iconButton favouriteIcon"
            data-testid="DetailPage_favBtn"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
              />
            </svg>
          </button>
        </h1>
        <h3>
          Pangasinan, Philippines
        </h3>
      </div>
      <div
        class="temperatureBlock"
      >
        <p
          class="temperature"
        >
          24
           ℃
        </p>
        <p>
          Moderate or heavy rain shower
        </p>
      </div>
    </div>
    <div
      class="lowerSection"
    >
      <div
        class="waRoot"
      >
        <div
          class="waTile"
        >
          <h3 />
          <p>
            Wind Speed
          </p>
        </div>
        <div
          class="waTile"
        >
          <h3>
            1006
          </h3>
          <p>
            Pressure
          </p>
        </div>
        <div
          class="waTile"
        >
          <h3>
            0.9
          </h3>
          <p>
            Precipitation
          </p>
        </div>
        <div
          class="waTile"
        >
          <h3>
            89
          </h3>
          <p>
            Humidity
          </p>
        </div>
        <div
          class="waTile"
        >
          <h3>
            37
          </h3>
          <p>
            Cloud Cover
          </p>
        </div>
        <div
          class="waTile"
        >
          <h3>
            9
          </h3>
          <p>
            Visibility
          </p>
        </div>
      </div>
      <div />
      <div>
        <div
          class="root"
        >
          <textarea
            placeholder="add note ..."
            rows="2"
          />
          <button>
            Add
          </button>
        </div>
        <div />
      </div>
    </div>
  </section>
</div>
`;
